README.md

```diff
@@ -1,32 +1,65 @@
+### Recovery
+Recovery consists of three steps:
+The first step is to recover the catalog. This is a tricky step as the catalog information is also contained in the data table files. We are still developing ideas for recovering the catalog. Currently, we use the existing log recovery for restoring the catalog.
+The second step is to recover all the data tables. The files are interpreted and dumped directly into the DataTables. For var-length datas, we reconstruct new pointers to the var-length content and replace the deprecated pointers.
+The third step is to recover the index. One solution would be to re-insert all the tuples into the index one by one and rebuild the index from the start. The other solution would be to store the primary index as well, so we can load the index directly. This approach, however, requires careful handling with pointers.
 
-   Yes, our goal is to have this new system to support autonomous operation and optimization as a first-class design principle.
 
-5. **Will the new system still be PostgreSQL compatiable?**
-   
-   Yes. The DBMS supports the PostgresSQL network protocol (both simple and extended) and emulates PostgresSQL's catalog layout. We are working on support for PL/pgSQL UDFs.
+## Design Rationale
+On a high level, we seperate the recovery of catalog tables from recovery from data tables. For catalog information and catalog table data and indices, we make use of the existing write-ahead logs to recover those portions. For all other user created data tables, we recover the content of the data tables with our checkpoint mechanism. This approach simplifies the implementation of checkpointing to stay in the scope of user-defined data-table. It also provides us with the assumption that catalog tables are fully recovered before we recover from the checkpoint. We believe this approach will still take the advantage of using checkpoint, since in a larger database the majority of data that requires recovery will lie within user defined data tables.
+
+Another major design decision we made is to deploy our own file format instead of using the existing arrow serializer. The reasoning is that we believe the current arrow format includes a lot of unnecessary information not needed for recovery. As we already performed all catalog recovery through the existing recovery from logging, recovery of the data tables’ content will require very little metadata about table columns. Therefore, we have created our own file format for this part.
+
+Finally, we design the recovery of table index to be performed in a tuple by tuple fashion. After we recover the table content, we iterate over each tuple available in the data table, and update the corresponding table index as an insert operation. We believe this approach makes full use of the efficiency of our checkpoints by eliminating the cost of performing deletes on indexes, and at the same time minimizes the design cost of hard coding updates into table indexes.
+
+## Testing Plan
```

Comment:
Missing tests:
- Detecting a failed/incomplete checkpoint and reverting to logs and/or previous checkpoint.
- Recovery when there is DML in the log after the most recent checkpoint (appears to be in-scope for your project)
- Recovery when there is DDL and DML in the log after the most recent checkpoitn (appears to be out-of-scope for your project, but should probably be at least documented as an important, but missing, test).